manpages/man/tiledb_subarray.1
.TH tiledb_subarray 1 "13 October 2015" "Version 0.1" "TileDB programs"
 
.SH NAME
tiledb_subarray - creates a new array with the same schema as the input array
(or including a subset of the attributes in a potentially different order), 
conataining only the cells that lie in the input range (i.e., hyper-rectangle). 

.SH SYNOPSIS
.B tiledb_subarray 
.BI "-A " "array_name,array_name_sub " "-r " "range "
.br
[
.BI "-w " "workspace,workspace_sub "
] [
.BI "-g " "group,group_sub "
] [
.BI "-a " "attribute_name_1,..."
] 

.SH DESCRIPTION
.B tiledb_subarray
creates a new array with the same schema as the input array (or including a 
subset of the attributes in a potentially different order), 
conataining only the cells that lie in the input range. The range must be a 
hyper-rectangle that is completely contained in the dimension space. It is also
given as a sequence of [low,high] pairs across each dimension. 
On error, it returns \fB-1\fR. Moreover, if it is compiled in \fBVERBOSE\fR 
mode, it prints comprehensive error messages on \fIstderr\fR. On success, it 
returns \fB0\fR. 

The following conventions apply to the \fBSYNOPSIS\fR section:

.TS
tab (@);
c lx .
\fBbold text\fR @ type exactly as shown
\fIitalic text\fR @ replace with appropriate argument
[\fB\-a \fIarg\fR]@ any or all options within [ ] are optional
.TE

.SH OPTIONS
.TP
.BI "-w" " workspace,workspace_sub" "\fR, \
" \fB --workspace=\fIworkspace,workspace_sub\fR  
The path to the workspace folder where the input array is stored, followed by
the workspace of the subarray result. If no workspace is provided, the current 
working directory is set as the workspace by default for both the input
and result arrays. If one workspace is provided, then this corresponds to the
workspace of the input array, but it will also be the workspace of the output
array as well.

.TP
.BI "-g" " group,group_sub" "\fR, " \fB --group=\fIgroup,group_sub\fR  
This is a directory inside the workspace where the input array is stored, 
followed by the directory (inside its respective workspace) where the
output subarray will be stored. Note that any group path provided is converted 
into an absolute path \fBrelative to its corresponding workspace directory\fR, 
i.e., regarding all home ("\\~/"), current ("./") and root ("/") as its 
workspace directory. If the group is not provided, the corresponding workspaces 
are set as the default groups for both input and output arrays. If one group is 
provided, then this corresponds to the input array, whereas the group of the 
output array is simply its workspace by default.

.TP
.BI "-A" " array_name,array_name_sub" "\fR, \
" \fB --array-name=\fIarray_name,array_name_sub\fR  
The name of the array the subarray will be applied on, followed by the name
of the result. The input array must be already defined. The name of the 
output array must be in POSIX format. 

.TP
.BI "-r" " range" "\fR, " \fB --range=\fIrange\fR  
The range given as a sequence of [low,high] bounds across each dimension.
Each range bound must be a real number. 

.TP
.BI "-a" " attribute_name_1,..." "\fR, " \
\fB --attribute-names=\fIattribute_name_1,...\fR  
The attributes to be included in the array schema of the output array. Any order
is allowed. To export \fBall\fR attributes, omit the option.

.SH EXAMPLES
Assuming that the array has dimensions \fIdim1\fR,\fIdim2\fR with domains
[0,100] and [0,200], respectively, and attributes 
\fIattr1\fR,\fIattr2\fR,\fIattr3\fR
.TP
\fBtiledb_subarray\fR \\ 
    \fB-w \fRmy_workspace/,my_workspace_sub/ \\
    \fB-g \fRmy_group/,my_group_sub/ \\
    \fB-A \fRmy_array,my_array_sub \\
    \fB-r \fR10,20,100,150 

Creates a new array \fImy_array_sub\fR with exactly the same schema as
\fImy_array\fR, but only containing the cells with \fIdim1\fR in 
\fB[10,20]\fR and \fIdim2\fR in \fB[100,150]\fR. The input array is
stored in \fImy_workspace/my_group/my_array\fR, whereas the output
array is stored in \fImy_workspace_sub/my_group_sub/my_array_sub\fR.

.TP
\fBtiledb_subarray\fR \\ 
    \fB-w \fRmy_workspace/,my_workspace_sub/ \\
    \fB-g \fRmy_group/,my_group_sub/ \\
    \fB-A \fRmy_array,my_array_sub \\
    \fB-r \fR10,20,100,150 \\ 
    \fB-a \fRattr1,attr2 

Same as the first example, but now \fImy_array_sub\fR contains only two
attributes, namely \fIattr1\fR and \fIattr2\fR.

.TP
\fBtiledb_subarray\fR \\ 
    \fB-w \fRmy_workspace/ \\
    \fB-g \fRmy_group/ \\
    \fB-A \fRmy_array,my_array_sub \\
    \fB-r \fR10,20,100,150 \\ 
    \fB-a \fRattr2,attr1

Same as the second example, but \fIattr1\fR and \fIattr2\fR appear
in a different order in \fImy_array_sub\fR. Also the output array
is stored in the same workspace and group as the input array,
namely \fImy_workspace/my_group/my_array\fR.

.SH EXIT STATUS
.TP 
\fB0\fR for success and \fB-1\fR for error.

.SH SEE ALSO
.BR "tiledb_array_define" "(1) 

.SH AUTHOR
Stavros Papadopoulos <http://people.csail.mit.edu/stavrosp>, Copyright (c) 2015


tiledb_cmd/src/tiledb_subarray.cc
/**
 * @file   tiledb_subarray.cc
 * @author Stavros Papadopoulos <stavrosp@csail.mit.edu>
 *
 * @section LICENSE
 *
 * The MIT License
 * 
 * @copyright Copyright (c) 2015 Stavros Papadopoulos <stavrosp@csail.mit.edu>
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 * 
 * @section DESCRIPTION
 *
 * Implements command "tiledb_subarray".
 */

#include "csv_line.h"
#include "special_values.h"
#include "tiledb.h"
#include "utils.h"
#include <getopt.h>
#include <iostream>
#include <string>

#define PRINT_ERROR(x) std::cerr << "[TileDB] Error: " << x << ".\n"
#define PRINT(x) std::cout << "[TileDB] " << x << ".\n"

/** 
 * Parses the command options. It returns 0 on success. On error, it prints
 * a message on stderr and returns -1.
 */                                                            
int parse_options(
    int argc, 
    char** argv, 
    std::string& workspace,
    std::string& workspace_sub,
    std::string& group,
    std::string& group_sub,
    std::string& array_name, 
    std::string& array_name_sub, 
    double*& range,
    int& range_size,
    std::vector<std::string>& attribute_names) {
  // Initialization
  workspace = "";
  workspace_sub = "";
  group = "";
  group_sub = "";
  array_name = "";
  array_name_sub = "";
  std::string workspaces_str, groups_str, array_names_str;
  std::string range_str("");
  std::string attribute_names_str("");
  attribute_names.clear();

  // Auxiliary variable
  CSVLine temp;

  // *************** //
  // Parse options * //
  // *************** // 

  // Define long options
  static struct option long_options[] = 
  {
    {"attribute-names",1,0,'a'},
    {"array-name",1,0,'A'},
    {"group",1,0,'g'},
    {"range",1,0,'r'},
    {"workspace",1,0,'w'},
    {0,0,0,0},
  };
  // Define short options
  const char* short_options = "a:A:g:r:w:";

  // Get options
  int c;
  int option_num = 0;
  while((c=getopt_long(argc, argv, short_options, long_options, NULL)) >= 0) {
    ++option_num;
    switch(c) {
      case 'a':
        if(attribute_names_str != "") {
          PRINT_ERROR("More than one attribute name lists provided");
          return -1;
        }
        attribute_names_str = optarg;
        break;
      case 'A':
        if(array_names_str != "") {
          PRINT_ERROR("More than one array name lists provided");
          return -1;
        }
        array_names_str = optarg;
        break;
      case 'g':
        if(groups_str != "") {
          PRINT_ERROR("More than one group lists provided");
          return -1;
        }
        groups_str = optarg;
        break;
      case 'r':
        if(range_str != "") {
          PRINT_ERROR("More than one ranges provided");
          return -1;
        }
        range_str = optarg;
        break;
      case 'w':
        if(workspaces_str != "") {
          PRINT_ERROR("More than one workspace lists provided");
          return -1;
        }
        workspaces_str = optarg;
        break; 
      default:
        return -1;
    }
  }

  // ******************* //
  // Check correctness * //
  // ******************* //

  // Check number of arguments
  if((argc-1) != 2*option_num) {
    PRINT_ERROR("Arguments-options mismatch");
    return -1;
  }
  //----- array names ----- //
  if(array_names_str == "") {
    PRINT_ERROR("Array names not provided");
    return -1;
  }
  temp = &array_names_str[0];
  if(temp.val_num() != 2) {
    PRINT_ERROR("Invalid number of array names");
    return -1;
  }
  temp >> array_name;
  temp >> array_name_sub;
  // ----- workspaces ----- //
  temp.clear();
  temp = &workspaces_str[0];
  if(temp.val_num() > 2) {
    PRINT_ERROR("Invalid number of workspaces");
    return -1;
  }
  temp >> workspace; 
  temp >> workspace_sub; 
  if(workspace_sub == "")
    workspace_sub = workspace;
  // ----- groups ----- //
  temp.clear();
  temp = &groups_str[0];
  if(temp.val_num() > 2) {
    PRINT_ERROR("Invalid number of groups");
    return -1;
  }
  temp >> group; 
  temp >> group_sub;
 // ----- attribute names ----- //
  if(attribute_names_str != "")
    attribute_names = CSVLine(&attribute_names_str[0]).values_str_vec();
  // ----- range ----- //
  std::vector<std::string> range_str_vec;
  range_str_vec = CSVLine(&range_str[0]).values_str_vec();
  range_size = int(range_str_vec.size());
  if(range_size == 0) {
    PRINT_ERROR("Range not provided");
    return -1;
  } else {
    range = new double[range_size];
    for(int i=0; i<range_size; ++i) {
      if(!is_real(range_str_vec[i].c_str())) {
        PRINT_ERROR("The range bounds must be real numbers");
        delete [] range;
        return -1;
      }
      sscanf(range_str_vec[i].c_str(), "%lf", &range[i]); 
    }
  }   

  // Success
  return 0;
}

int main(int argc, char** argv) {
  // Parse command line
  std::string workspace;
  std::string workspace_sub;
  std::string group;
  std::string group_sub;
  std::string array_name;
  std::string array_name_sub;
  double* range = NULL;
  int range_size = 0;
  const char** attribute_names_c_str;
  int attribute_names_num = 0;
  std::vector<std::string> attribute_names;
 
  if(parse_options(
         argc, argv, workspace, workspace_sub, group, group_sub,
         array_name, array_name_sub, range, range_size,
         attribute_names))
    return -1;

  // Initialize TileDB
  TileDB_CTX* tiledb_ctx;
  if(tiledb_ctx_init(tiledb_ctx))
    return -1;

  // Get vector for attribute names
  attribute_names_num = int(attribute_names.size());
  if(attribute_names_num == 0) {
    attribute_names_c_str = NULL;
  } else {
    attribute_names_c_str = new const char*[attribute_names_num];
    for(int i=0; i<attribute_names_num; ++i) 
      attribute_names_c_str[i] = attribute_names[i].c_str();
  }

  // Subarray
  if(tiledb_subarray(
         tiledb_ctx, 
         workspace.c_str(), workspace_sub.c_str(),
         group.c_str(), group_sub.c_str(),
         array_name.c_str(), array_name_sub.c_str(),
         range, range_size,
         attribute_names_c_str, attribute_names_num)) {
    // Error
    if(attribute_names_c_str != NULL) 
      delete [] attribute_names_c_str;
    if(range != NULL)
      delete [] range;
    tiledb_ctx_finalize(tiledb_ctx);
    return -1;
  }

  // Clean up  
  if(attribute_names_c_str != NULL) 
    delete [] attribute_names_c_str;
  if(range != NULL)
    delete [] range;

  // Finalize TileDB
  if(tiledb_ctx_finalize(tiledb_ctx))
    return -1;

  // Success
  PRINT("Program executed successfully");
  
  return 0;
}
